manim/scene/scene.py

```diff
@@ -628,19 +629,24 @@ def get_restructured_mobject_list(self, mobjects: list, to_remove: list):
             The list of mobjects with the mobjects to remove removed.
         """
 
-        new_mobjects = []
-
         def add_safe_mobjects_from_list(list_to_examine, set_to_remove):
+            new_mobjects = []
             for mob in list_to_examine:
                 if mob in set_to_remove:
                     continue
                 intersect = set_to_remove.intersection(mob.get_family())
                 if intersect:
-                    add_safe_mobjects_from_list(mob.submobjects, intersect)
+                    newmob = mob.copy()
```

Comment:
**Question :** I'm a little reluctant about using a copy here, both for performance reasons and most importantly, it can probably break some things (the references of the submobjects will get change).

I don't understand, why using a `copy` here ?

Reply:
Using a copy here was my attempt to retain the current behaviour of remove. Prior to the commits in this PR, running remove would not affect the original mobject that was added to the scene. But with this change, if I didn't use `.copy()`, it would produce an error in code like the second snipped from [this comment](https://github.com/ManimCommunity/manim/issues/3027#issuecomment-1321851304), which has a reference to the submobject after it gets removed from the scene.

Without using `copy()` it may be difficult to solve the problem.

Reply:
It's been more than 1 year, but...

IMO the current implementation using `Mobject.get_family()` is more convoluted than it should, given that one's plan after filtering the main list is to filter each mobject's own `submobjects` list, not much to do with the whole family itself.

It's much more straightforward to directly go ahead and recurse over the `submobjects`, without going out of your way and getting _all_ the family.

Take a look at this code. It has no copying, and yet it does not have the removal bug in the link you mentioned. Maybe it could work for you :eyes: 

```py
    def get_restructured_mobject_list(self, mobjects: list, to_remove: list):
        def filter_mobjects(list_to_examine, set_to_remove):
            new_mobjects = []
            for mob in list_to_examine:
                if mob in set_to_remove:
                    continue
                mob.submobjects = filter_mobjects(mob.submobjects, set_to_remove)
                new_mobjects.append(mob)
            return new_mobjects

        return filter_mobjects(mobjects, set(to_remove))